### PR TITLE
libelf: call autoreconf + tiny refactor of CMakeLists + allow shared libs

### PR DESCRIPTION
**Pull request by SpaceIm**
Specify library name and version:  **libelf/all**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

closes https://github.com/conan-io/conan-center-index/issues/2795

### CODE REVIEW

File: recipes/libelf/all/conanfile.py [hunk 1]
```diff
@@ -28,17 +29,20 @@ def _build_subfolder(self):
...
     def configure(self):
         if self.options.shared:
             del self.options.fPIC
         del self.settings.compiler.libcxx
         del self.settings.compiler.cppstd
-        if self.settings.os != "Linux":
-            if self.options.shared:
-                raise ConanInvalidConfiguration("libelf can not be built as shared library on non linux platforms")
+        if self.options.shared and self.settings.os not in ["Linux", "Windows"]:
+            raise ConanInvalidConfiguration("libelf can not be built as shared library on non linux or windows platforms")
```
Comment (AndreyMlashkin) [inline recipes/libelf/all/conanfile.py:41]: Is there any particular reason why libelf can not be build as a shared lib on MacOs?

**Review by AndreyMlashkin — APPROVED** [review]

File: recipes/libelf/all/conanfile.py (hunk 1 above)
Comment (SpaceIm) [reply recipes/libelf/all/conanfile.py:41]: Because their autoconf file is not very portable.

File: recipes/libelf/all/CMakeLists.txt [hunk 2]
```diff
@@ -1,7 +1,33 @@
...
+
+file(GLOB_RECURSE SOURCES "source_subfolder/lib/*.c")
+file(GLOB_RECURSE HEADERS "source_subfolder/lib/*.h")
+
+add_library(${PROJECT_NAME} ${SOURCES})
+set_target_properties(${PROJECT_NAME} PROPERTIES PUBLIC_HEADER "${HEADERS}")
+target_include_directories(${PROJECT_NAME} SYSTEM PUBLIC "source_subfolder/lib")
+target_compile_definitions(${PROJECT_NAME} PUBLIC HAVE_MEMCMP=1 HAVE_MEMCPY=1 HAVE_MEMMOVE=1)
+
+if(MSVC AND BUILD_SHARED_LIBS)
```
Comment (madebr) [inline recipes/libelf/all/CMakeLists.txt:20]:
```suggestion
if(WIN32AND BUILD_SHARED_LIBS)
```

Comment (SpaceIm) [reply recipes/libelf/all/CMakeLists.txt:20]: WINDOWS_EXPORT_ALL_SYMBOLS is ignored if not MSVC or not SHARED, so it doesn't really matter.

Comment (madebr) [reply recipes/libelf/all/CMakeLists.txt:20]: I think this also works for mingw.
https://cmake.org/cmake/help/latest/prop_tgt/WINDOWS_EXPORT_ALL_SYMBOLS.html is talking about MS-compatible tools.

Comment (SpaceIm) [reply recipes/libelf/all/CMakeLists.txt:20]: MinGW is like others GNU tools: symbols are exported by default.

Comment (madebr) [reply recipes/libelf/all/CMakeLists.txt:20]: No that's not true. You also need to export symbols with a declaration.
http://www.mingw.org/wiki/sampledll

Comment (SpaceIm) [reply recipes/libelf/all/CMakeLists.txt:20]: you mean global variables? WINDOWS_EXPORT_ALL_SYMBOLS doesn't export those symbols, manual export is still required.

Comment (madebr) [reply recipes/libelf/all/CMakeLists.txt:20]: WINDOWS_EXPORT_ALL_SYMBOLS will export all non-static functions and classes.
For both msvc and mingw.
And indeed, manual export is required for global variables.

What I want to say is that it works the same for mingw and msvc.

Comment (SpaceIm) [reply recipes/libelf/all/CMakeLists.txt:20]: I've tested, and symbols are exported if MinGW and shared, without adding `WINDOWS_EXPORT_ALL_SYMBOLS`.

**Review by jgsogo — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Review by danimtb — APPROVED** [review]

**Pull request merged**